File: windows/src/LanStash.Infrastructure/DsmConnectionResolver.cs
using LanStash.Domain;

namespace LanStash.Infrastructure;

public sealed record DiscoveredConnection(
    NasProfile Profile,
    IReadOnlyDictionary<string, ApiCapability> Capabilities);

/// <summary>
/// 登录前只使用不含凭据的能力发现探测连接候选。
/// 找到可信连接后，调用方才可以提交账号、密码和验证码。
/// </summary>
public sealed class DsmConnectionResolver(
    IDsmApiClient api,
    DsmQuickConnectResolver quickConnect)
{
    public async Task<DiscoveredConnection> DiscoverAsync(
        NasProfile profile,
        Action<string>? updateStatus = null,
        CancellationToken cancellationToken = default)
    {
        var parsed = NasAddressParser.Parse(profile.Host, profile.Port);
        if (parsed.Kind == NasAddressKind.Direct)
        {
            updateStatus?.Invoke("正在连接 NAS…");
            var directProfile = profile with { Host = parsed.Host, Port = parsed.Port };
            return new(
                directProfile,
                await api.DiscoverAsync(directProfile, cancellationToken));
        }

        updateStatus?.Invoke("正在通过 QuickConnect 查找 NAS…");
        IReadOnlyList<QuickConnectEndpoint> endpoints;
        try
        {
            endpoints = await quickConnect.ResolveAsync(parsed.Host, cancellationToken);
        }
        catch (DsmException error) when (error.Message == "QuickConnect 没有提供可用的直接连接")
        {
            endpoints = [];
        }

        DsmException? lastDirectError = null;
        foreach (var endpoint in endpoints)
        {
            updateStatus?.Invoke(endpoint.Kind == QuickConnectEndpointKind.Local
                ? "正在尝试局域网连接…"
                : "正在尝试外网直接连接…");
            var connectionProfile = profile with
            {
                Host = endpoint.Host,
                Port = profile.Port ?? endpoint.Port,
            };
            try
            {
                return new(
                    connectionProfile,
                    await api.DiscoverAsync(connectionProfile, cancellationToken));
            }
            catch (DsmException error)
            {
                lastDirectError = error;
            }
        }

        updateStatus?.Invoke("正在建立 QuickConnect 安全中继…");
        try
        {
            var relay = await quickConnect.RequestRelayAsync(parsed.Host, cancellationToken);
            var relayProfile = profile with { Host = relay.Host, Port = relay.Port };
            return new(
                relayProfile,
                await api.DiscoverAsync(relayProfile, cancellationToken));
        }
        catch (DsmException error) when (
            error.Message == "QuickConnect 暂时无法建立中继连接" && lastDirectError is not null)
        {
            throw lastDirectError;
        }
    }
}

File: android/app/src/main/java/io/github/qwertyuiop1995/dsmnativeclient/network/DsmConnectionResolver.kt
package io.github.qwertyuiop1995.dsmnativeclient.network

import io.github.qwertyuiop1995.dsmnativeclient.domain.ApiCapability
import io.github.qwertyuiop1995.dsmnativeclient.domain.DsmFailure
import io.github.qwertyuiop1995.dsmnativeclient.domain.NasProfile
import kotlinx.coroutines.CancellationException

internal data class DiscoveredConnection(
    val profile: NasProfile,
    val capabilities: Map<String, ApiCapability>,
)

/**
 * 登录前只使用不含凭据的能力发现探测连接候选。
 * 找到可信连接后，调用方才可以提交账号、密码和验证码。
 */
internal class DsmConnectionResolver(
    private val api: DsmApiClient,
    private val quickConnect: DsmQuickConnectResolver = DsmQuickConnectResolver(),
) {
    suspend fun discover(
        profile: NasProfile,
        onStatus: (String) -> Unit = {},
    ): DiscoveredConnection {
        val parsed = NasAddressParser.parse(profile.address, profile.port)
        if (parsed.kind == NasAddressKind.DIRECT) {
            onStatus("正在连接 NAS…")
            val connectionProfile = profile.copy(
                address = "https://${parsed.host}",
                port = parsed.port,
            )
            return DiscoveredConnection(connectionProfile, api.discover(connectionProfile))
        }

        onStatus("正在通过 QuickConnect 查找 NAS…")
        val endpoints = try {
            quickConnect.resolve(parsed.host)
        } catch (error: DsmFailure) {
            if (error.message == "QuickConnect 没有提供可用的直接连接") {
                emptyList()
            } else {
                throw error
            }
        }
        var lastDirectFailure: DsmFailure? = null
        for (endpoint in endpoints) {
            onStatus(
                if (endpoint.kind == QuickConnectEndpointKind.LOCAL) {
                    "正在尝试局域网连接…"
                } else {
                    "正在尝试外网直接连接…"
                }
            )
            val connectionProfile = profile.copy(
                address = "https://${endpoint.host}",
                port = profile.port ?: endpoint.port,
            )
            try {
                return DiscoveredConnection(connectionProfile, api.discover(connectionProfile))
            } catch (error: CancellationException) {
                throw error
            } catch (error: DsmFailure) {
                lastDirectFailure = error
            }
        }

        onStatus("正在建立 QuickConnect 安全中继…")
        return try {
            val relay = quickConnect.requestRelay(parsed.host)
            val connectionProfile = profile.copy(
                address = "https://${relay.host}",
                port = relay.port,
            )
            DiscoveredConnection(connectionProfile, api.discover(connectionProfile))
        } catch (error: CancellationException) {
            throw error
        } catch (error: DsmFailure) {
            throw if (error.message == "QuickConnect 暂时无法建立中继连接") {
                lastDirectFailure ?: error
            } else {
                error
            }
        }
    }
}
